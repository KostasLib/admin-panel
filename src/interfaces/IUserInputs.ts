export interface IUSerInputs {
  name: string;
  email: string;
  phone: string;
  address: string;
  company: string;
}
