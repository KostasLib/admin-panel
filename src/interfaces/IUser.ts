import { IUSerInputs } from './IUserInputs';

export interface IUser extends IUSerInputs {
  id: string;
  photo?: string;
}
