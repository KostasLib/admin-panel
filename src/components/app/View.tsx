import React from 'react';
import { UserList } from '../user-list';
import { UserForm } from '../user-form';
import { IUSerInputs } from '../../interfaces/IUserInputs';
import { IUser } from '../../interfaces/IUser';
import s from './View.module.scss';

interface IProps {
  users: IUser[];
  selected?: string;
  onSelectUser: (id: string) => () => void;
  onSave: (newData: IUSerInputs) => void;
}

/** A stateless function component that handle's the App's presentation layer. */
export const View: React.FC<IProps> = props => {
  const { onSave, users, selected, onSelectUser } = props;

  return (
    <article className={s.panel}>
      <section className={s['squeezeable-section']}>
        <UserList onSelectUser={onSelectUser} users={users} selected={selected} />
      </section>
      <section className={s['form-section']}>
        <UserForm user={users.find(u => u.id === selected)} onSave={onSave} />
      </section>
    </article>
  );
};
