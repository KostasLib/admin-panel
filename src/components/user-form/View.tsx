import React from 'react';
import { IUSerInputs } from '../../interfaces/IUserInputs';
import { Input } from '../input';
import s from './UserForm.module.scss';

interface IProps {
  inputs: IUSerInputs;
  hasChanged: boolean;
  onChange: (name: keyof IUSerInputs) => (e: any) => void;
  onCancel: (e: React.FormEvent<HTMLFormElement>) => void;
  onSave: (e: React.FormEvent<HTMLFormElement>) => void;
}

/** A stateless function component that handles the presentation for the UserForm. */
export const View: React.FC<IProps> = props => {
  const {
    inputs: { name, email, phone, address, company },
    onChange,
    hasChanged,
    onCancel,
    onSave,
  } = props;

  return (
    <form onSubmit={onSave} onReset={onCancel}>
      <Input
        label="Name"
        type="text"
        name="name"
        placeholder="Enter name"
        title="Enter name"
        value={name}
        required
        maxLength={255}
        onChange={onChange('name')}
      />
      <Input
        label="Email address"
        type="email"
        name="email"
        placeholder="Enter email"
        title="Enter email"
        value={email}
        maxLength={255}
        onChange={onChange('email')}
      />
      <Input
        label="Phone"
        type="tel"
        name="phone"
        placeholder="Enter phone"
        title="Enter phone"
        value={phone}
        maxLength={255}
        pattern="^[^a-zA-Z]*$"
        onChange={onChange('phone')}
      />

      <Input
        label="Address"
        type="text"
        name="address"
        placeholder="Enter address"
        title="Enter address"
        value={address}
        maxLength={255}
        onChange={onChange('address')}
      />
      <Input
        label="Company"
        type="text"
        name="company"
        placeholder="Enter company"
        title="Enter company"
        value={company}
        maxLength={255}
        onChange={onChange('company')}
      />

      <div className={s.buttons}>
        {hasChanged && (
          <input className={s.cancel} type="reset" value="Cancel" title="Cancel" />
        )}
        <input
          className={s.save}
          type="submit"
          value="Save"
          title="Save changes"
          disabled={!hasChanged}
        />
      </div>
    </form>
  );
};
