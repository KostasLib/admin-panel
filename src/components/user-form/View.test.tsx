import React from 'react';
import { shallow } from 'enzyme';
import { View } from './View';
import { IUSerInputs } from '../../interfaces/IUserInputs';

describe("The user form's <View /> component", () => {
  const mockInputs: IUSerInputs = {
    name: '',
    email: '',
    address: '',
    company: '',
    phone: '',
  };

  it("renders the form's view with both buttons enabled when hasChanged is true", () => {
    const mockOnCancel = jest.fn();
    const mockOnSave = jest.fn();
    const mockOnChange = jest.fn();
    const component = shallow(
      <View
        inputs={mockInputs}
        onChange={mockOnChange}
        onCancel={mockOnCancel}
        onSave={mockOnSave}
        hasChanged={true}
      />,
    );

    // Snapshot
    expect(component).toMatchSnapshot();

    // Assert that buttons are not disabled.
    expect(component.find('input[type="reset"]').prop('disabled')).toBeFalsy();
    expect(component.find('input[type="submit"]').prop('disabled')).toBeFalsy();
  });

  it('renders only the submit button as disabled when no input has changed', () => {
    const mockOnCancel = jest.fn();
    const mockOnSave = jest.fn();
    const mockOnChange = jest.fn();
    const component = shallow(
      <View
        inputs={mockInputs}
        onChange={mockOnChange}
        onCancel={mockOnCancel}
        onSave={mockOnSave}
        hasChanged={false}
      />,
    );

    // Snapshot
    expect(component).toMatchSnapshot();

    // Assert the submit button is disabled.
    expect(component.find('input[type="submit"]').prop('disabled')).toBe(true);
  });

  it('registers an onChange handler for each input field', () => {
    const mockOnCancel = jest.fn();
    const mockOnSave = jest.fn();
    const mockRegisterOnChange = jest.fn();
    const component = shallow(
      <View
        inputs={mockInputs}
        onChange={mockRegisterOnChange}
        onCancel={mockOnCancel}
        onSave={mockOnSave}
        hasChanged={false}
      />,
    );

    expect(mockRegisterOnChange).toHaveBeenCalledTimes(Object.keys(mockInputs).length);
  });

  it('runs event handlers when the input changes, on form submit and on form reset', () => {
    const mockOnCancel = jest.fn();
    const mockOnSave = jest.fn();
    const mockHandleChange = jest.fn();
    const mockOnChange = jest.fn(() => mockHandleChange);
    const component = shallow(
      <View
        inputs={mockInputs}
        onChange={mockOnChange}
        onCancel={mockOnCancel}
        onSave={mockOnSave}
        hasChanged={false}
      />,
    );

    // Test handleChange
    component
      .find('Input')
      .first()
      .simulate('change');
    expect(mockHandleChange).toHaveBeenCalled();

    // Test submit
    component.find('form').simulate('submit');
    expect(mockOnSave).toHaveBeenCalled();

    // Test reset
    component.find('form').simulate('reset');
    expect(mockOnCancel).toHaveBeenCalled();
  });
});
