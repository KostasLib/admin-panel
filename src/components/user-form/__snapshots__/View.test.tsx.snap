// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The user form's <View /> component renders only the submit button as disabled when no input has changed 1`] = `
<form
  onReset={[MockFunction]}
  onSubmit={[MockFunction]}
>
  <Input
    label="Name"
    maxLength={255}
    name="name"
    placeholder="Enter name"
    required={true}
    title="Enter name"
    type="text"
    value=""
  />
  <Input
    label="Email address"
    maxLength={255}
    name="email"
    placeholder="Enter email"
    title="Enter email"
    type="email"
    value=""
  />
  <Input
    label="Phone"
    maxLength={255}
    name="phone"
    pattern="^[^a-zA-Z]*$"
    placeholder="Enter phone"
    title="Enter phone"
    type="tel"
    value=""
  />
  <Input
    label="Address"
    maxLength={255}
    name="address"
    placeholder="Enter address"
    title="Enter address"
    type="text"
    value=""
  />
  <Input
    label="Company"
    maxLength={255}
    name="company"
    placeholder="Enter company"
    title="Enter company"
    type="text"
    value=""
  />
  <div
    className="buttons"
  >
    <input
      className="save"
      disabled={true}
      title="Save changes"
      type="submit"
      value="Save"
    />
  </div>
</form>
`;

exports[`The user form's <View /> component renders the form's view with both buttons enabled when hasChanged is true 1`] = `
<form
  onReset={[MockFunction]}
  onSubmit={[MockFunction]}
>
  <Input
    label="Name"
    maxLength={255}
    name="name"
    placeholder="Enter name"
    required={true}
    title="Enter name"
    type="text"
    value=""
  />
  <Input
    label="Email address"
    maxLength={255}
    name="email"
    placeholder="Enter email"
    title="Enter email"
    type="email"
    value=""
  />
  <Input
    label="Phone"
    maxLength={255}
    name="phone"
    pattern="^[^a-zA-Z]*$"
    placeholder="Enter phone"
    title="Enter phone"
    type="tel"
    value=""
  />
  <Input
    label="Address"
    maxLength={255}
    name="address"
    placeholder="Enter address"
    title="Enter address"
    type="text"
    value=""
  />
  <Input
    label="Company"
    maxLength={255}
    name="company"
    placeholder="Enter company"
    title="Enter company"
    type="text"
    value=""
  />
  <div
    className="buttons"
  >
    <input
      className="cancel"
      title="Cancel"
      type="reset"
      value="Cancel"
    />
    <input
      className="save"
      disabled={false}
      title="Save changes"
      type="submit"
      value="Save"
    />
  </div>
</form>
`;
